feat(billing services): implement billing services client-side

jniles opened the pull request:
This pull requests implements the billing services client.  While doing this, we demonstrate the power and usefulness of UI-Router.
#### Screenshots

**Empty Billing Services Grid**
![billingservicesempty](https://cloud.githubusercontent.com/assets/896472/14104188/afd97ae2-f59b-11e5-9bc6-c640f6103aa4.png)

**Creating a Billing Service**
![billingservicescreate](https://cloud.githubusercontent.com/assets/896472/14104191/b0206df8-f59b-11e5-917e-9fb636861052.png)

**Updating A Billing Service**
![billingservicesupdate](https://cloud.githubusercontent.com/assets/896472/14104187/afca0aa8-f59b-11e5-9828-3cf3a68d7903.png)

**Deleting A Billing Service**
![billingservicesdelete](https://cloud.githubusercontent.com/assets/896472/14104189/aff0893a-f59b-11e5-89ec-b09a0c7eb02a.png)

**Billing Service Grid Populated**
![billingservicescreated](https://cloud.githubusercontent.com/assets/896472/14104190/b0027ad2-f59b-11e5-8a13-7f71cb46d008.png)
#### Summary

The billing services module went through many transformations (see the commits) before settling on the best practices of using modals + ui-router.  This is a pattern that I believe is robust for simple modules (e.g. projects, enterprises, debtor groups).  However, it will not be scalable or appropriate to larger modules.

This PR demonstrates some significant advantages of ui-router.  Here are some highlights:
1. We can "deep-link" complex states.  For example, by going to the URL `/biling_services/:id/update`, from anywhere in the application, you will automatically arrive at a "update" (modal) form for the billing service of that id.  Clicking the "back" button on the form will take you to the billing services module _without reloading the view_.  This is an incredibly powerful feature.
2. The controllers are cleanly separated into `update`, `create`, and `delete`.  There is no longer any need to implement a check like
   `js
   var promise = (action === 'create') ? BillingServices.create(data) : BillingServices.update(id, data);
   `
   These states are no longer entangled.  The `UpdateController` does nothing but think about what is necessary to update.  The `CreateController` only thinks about creating.  They can both use the same form to accomplish this task, but the logic is cleanly separated.
3. We can now selectively refresh only part of the view, using `<ui-view>`.  While this isn't well demonstrated in this PR, it will help more complex modules.

There are still some significant issues with the billing services UI.  See https://github.com/IMA-WorldHealth/bhima-2.X/commit/bb1c575dd58ed312f7bffe56bac3713dddcef945 for a discussion of them.  I will be making a review issue to track changes to the UI.

Closes #51.

jniles committed:
package(bower): update depends and add ui-router

jniles committed:
deps: update bower modules

jniles committed:
feat: ui-router for routing

This commit adds in ui-router to the bower dependencies, removes
ngRoute, and adjusts the code to work with ui-router.

jniles committed:
feat: inital billing services implementation

This implements the intial structure of the billing services page.  As
of this commit, billing service creation works with proper form
validation.  A skeleton UI grid can also be found on the billing
services page.

The struture of billing services looks like this:
 - billing_services/
   - billing_services.html # the abstract state template with title bar)
   - billing_services.list.html # the default grid for the page
   - billing_service.js  # the BillingServicesController
   - create/
     - create.html  # the creation form
     - create.js    # the BillingServicesCreateController
   - update/
     - update.html  # the update form
     - update.js    # the BillingServicesUpdateController

The routes should look like this:
 - `/billing_services` - lists all billing services in a UI grid
 - `/billing_services/create` - brings up a creation form
 - `/billing_services/:id` - brings up an update form

 TODOs
  - implement the `?detailed=1` query string on the server +
    integration tests
  - implement the UPDATE state

jniles committed:
chore(billing services): implement ?detailed=1

This commit implements the `?detailed=1` query string route for the GET
/billing_services API endpoint.  An integraiton test has been added to
prove it works.

Other improvements:
 1. The billing_services integration tests now use before() instead of
 beforeEach() to authenticate the agent.
 2. The API uses proper error codes.
 3. Lodash ES6-stype template strings for descriptions while we wait
 approval to use actual ES6 template strings.

jniles committed:
chore(billing services): client structure changes

This commit changes the structure of the client to allow the abstract
state better control over child states, and access to the $state
service.  This will make it possible in the future to hide/show buttons
in the headercrumb based on the $state.current.

Importantly, this commit introduces the pattern of
 - /list
   - list.js
   - list.html
for default lists.  In the future, we might want to change this to
 - /index
   - index.js
   - index.html
if it makes more sense (such as most pages not using lists for their
interface).

Finally, the grid now has loading indicators and error indicators, which
should ease user experience in low-bandwidth environments.

jniles committed:
chore(billing services): create/update working

This commit moves the CREATE and UPDATE controls to using the same form
template, now located at `/billing_services/form.html`.  The methods
both work as expected.

jniles committed:
feat(filters): percentage filter implementation

This commit adds in a simply percentage filter for adding percent (%)
signs to numbers.  If the input is not a number, the percent filter does
not do anything

jniles committed:
chore: i18n, comments, and error handling

This commit cleans up the billing services controllers and forms with better
error handling, translations, and comments.  The only thing remaining
are tests.

jniles committed:
test(billing services): implement e2e tests

This commit implements end-to-end tests for billing services.  So far,
only the create and update tests are supported. Deletion will be
targetted next once we have a modal component to confirm deletes.

jniles committed:
feat: initial confirm modal implementation

This commit adds in a modal replacement for $window.confirm() for better
security.  See #150 for reference.

The modal is available in the ModalService, which also contains some
TODOs for later development.  The service exposes a single method,
`confirm()` that will resolve with a boolean value, depending on what
the user click - "true" for accept and "false" for cancel.

jniles committed:
test(billing services): add delete e2e test

This commit adds a simple end-to-end test to check that delete works on
the billing service module.

jniles committed:
feat(billing services): delete modal + deep link

This commit adds the ability to deep link deletes via ui-router's state.
Delete now operates like update, using `/billing_services/:id/delete` to
open a generic delete modal to perform the delete request.

This commit completes the functionality of the billing_services module.
However, it needs polishing.  A review issue will be made to mention the
following things specfically:
 1. Bad Delete Modal UI - this should be improved as we use it in
 multiple places.
 2. The "create" button is in a terribly chosen place - should this be a
 dropdown?
 3. (Server) error handling needs to be decided on and adopted by the
 whole team uniformly.
 4. Better alert system (scroll to and highlight) for ui-grid on create
 and update.

 This commit officially closes #206.

jniles committed:
chore(hound): style fixes

This commit brings the PR up to date with style fixes for violations
found by houndci's bot.

sfount commented on client/src/js/services/AccountService.js:
```diff
@@ -6,30 +6,64 @@ AccountService.$inject = ['$http', 'util'];
+      .then(function (accounts) {
+
+        // label the accounts with a nice human readable label
+        if (!id) {
```
To make sure I understand correctly - Is the purpose of this statement to only apply the label if we receive an array of accounts? 

Given the id's use here what would happen if I did not want to specify an account but did want to provide options? 
Something like this: 

``` js
Accounts.read(null, { options... })...
```

This could have implications on the branching structure (update: tested this and it doesn't break), potentially something like `Angular.isArray` would be safer/ giver more reliable results. 

(An id of 0 would also break this condition).

jniles replied:
Thanks for your feedback.  I agree that `Angular.isArray` is clearer to read, and probably safer.  An id of 0 shouldn't ever exist in our database...but I'll refactor this for clarity's sake.

jniles committed:
chroe(accounts): clean up account service

This commit cleans up the account service's code by moving a stricter
angular.isArray check instead of a boolean id check.

jniles committed:
fix: merge issues

This commit fixes the following issues with the latest merge from
master:
 1. reference groups - broken routes in app.js
 2. accounts - the API changed, breaking billing_services
 3. bh-date-editor -  the patient invoice module no longer has two
 "date" labels
 4. patient registration - the registration page now routes to
 patients/edit.
